binary_search_tree/tree.py

```diff
@@ -7,51 +7,159 @@ def __init__(self, key, val = None):
+
+        if not current:
+            current = self.root
+
+        if current.key >= key:
+            if current.left:
+                self.add(key, value, current.left)
+            else:
+                current.left = TreeNode(key, value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+        if current.key < key:
+            if current.right:
+                self.add(key, value, current.right)
+            else:
+                current.right = TreeNode(key, value)
+
+    # Time Complexity: O ( log n )
+    # Space Complexity: O ( log n )
     def find(self, key):
-        pass
+        if not self.root:
+            return None
+
+        current = self.root
+        while current.key != key:
+            if current.key > key and current.left:
+                current = current.left
+            elif current.key < key and current.right:
+                current = current.right
+            elif not (current.left and current.right):
+                return None
 
-    # Time Complexity: 
-    # Space Complexity: 
+        return current.value
+
+    # Time Complexity: O ( log n )
+    # Space Complexity: O ( n )
+    # left, center, right
     def inorder(self):
-        pass
+        if not self.root:
+            return []
+
+        return self.inorder_help(self.root)
+
+    def inorder_help(self, current):
+        ret_list = []
+
+        if current.left:
+            ret_list += self.inorder_help(current.left)
 
-    # Time Complexity: 
-    # Space Complexity:     
+        ret_list.append(current.formatted())
+
+        if current.right:
+            ret_list += self.inorder_help(current.right)
+
+        return ret_list
+
+    # Time Complexity: O ( log n )
+    # Space Complexity: O ( n )
+    # center, left, right     
     def preorder(self):
-        pass
+        if not self.root:
+            return []
+
+        return self.preorder_help(self.root)
+
+    def preorder_help(self, current):
+        ret_list = []
+
+        ret_list.append(current.formatted())
+
+        if current.left:
+            ret_list += self.preorder_help(current.left)
+
+        if current.right:
+            ret_list += self.preorder_help(current.right)
 
-    # Time Complexity: 
-    # Space Complexity:     
+        return ret_list
+
+    # Time Complexity: O ( log n )
+    # Space Complexity: O ( n )
+    # left, right, center 
     def postorder(self):
-        pass
+        if not self.root:
+            return []
+
+        return self.postorder_help(self.root)
+
+    def postorder_help(self, current):
+        ret_list = []
+
+        if current.left:
+            ret_list += self.postorder_help(current.left)
 
-    # Time Complexity: 
-    # Space Complexity:     
+        if current.right:
+            ret_list += self.postorder_help(current.right)
+
+        ret_list.append(current.formatted())
+
+        return ret_list
+
+    # Time Complexity: O ( n )
+    # Space Complexity: O ( n )
     def height(self):
```

Comment:
👍  The space complexity is O(h) or O(log n) for a balanced tree because the call stack won't get any larger than the height of the tree.